lib/srv/db/common/auth.go

```diff
@@ -505,6 +516,56 @@ func (a *dbAuth) Close() error {
 	return a.cfg.Clients.Close()
 }
 
+// generateCloudSQLClientCertificate returns a new client certificate with RSA key generated
+// using Google's GenerateEphemeralCertRequest sqladmin API. Client certificates as required
+// when enabling SSL in Cloud SQL.
+func (a *dbAuth) generateCloudSQLClientCertificate(ctx context.Context, sessionCtx *Session) (tls.Certificate, error) {
+	svc, err := a.cfg.Clients.GetGCPSQLAdminClient(ctx)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+
+	// generate RSA private key, x509 encoded public key,
+	// and append to certificate request
+	pkey, err := rsa.GenerateKey(rand.Reader, constants.RSAKeySize)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	pkix, err := x509.MarshalPKIXPublicKey(pkey.Public())
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	gcp := sessionCtx.Database.GetGCP()
+	req := svc.Connect.GenerateEphemeralCert(gcp.ProjectID, gcp.InstanceID, &sqladmin.GenerateEphemeralCertRequest{
+		PublicKey: string(pem.EncodeToMemory(&pem.Block{Bytes: pkix, Type: "RSA PUBLIC KEY"})),
+	})
+
+	// make request to generate certificate
+	resp, err := req.Do()
+	if err != nil {
+		return tls.Certificate{}, trace.Errorf("requesting client certificate: %w", err)
+	}
+
+	// PEM decode and parse returned x509 certificate
+	bl, _ := pem.Decode([]byte(resp.EphemeralCert.Cert))
+	if bl == nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	cert, err := x509.ParseCertificate(bl.Bytes)
+	if err != nil {
+		return tls.Certificate{}, trace.Errorf("parsing certificate for instance %q: %w", gcp.InstanceID, err)
+	}
+
+	// create client certificate
+	tlscert := tls.Certificate{
+		Certificate: [][]byte{cert.Raw},
+		PrivateKey:  pkey,
+		Leaf:        cert,
+	}
```

Comment:
I would use [tls.X509KeyPair](https://pkg.go.dev/crypto/tls#X509KeyPair) method instead of constructing tls.Certificate manually.